site/reference/annotations/apply-time-mutation/README.md

```diff
@@ -0,0 +1,235 @@
+substitutions to make at apply time using dependencies as input.
+
+### Schema
+
+The annotation value requires a YAML string.
+
+- ([]<a href="#substitution">Substitution</a>), required
+
+    A list of one or more substitutions to perform on the annotated object.
+
+#### Substitution
+
+A substitution is a modification to a specific target field.
+
+-   **sourceRef**: (<a href="#ObjectReference">ObjectReference</a>), required
+
+    Reference to the source resource, the dependency for this substitution.
+
+-   **sourcePath**: (string), required
+
+    The source resource field to read from as input, specified by
+    <a href="#jsonpath-syntax">JSONPath</a>.
+
+-   **targetPath**: (string), required
+
+    The target resource field to write to as output, specified by
+    <a href="#jsonpath-syntax">JSONPath</a>.
+
+-   **token**: (string)
+
+    The substring to replace in the target resource field.
+
+    If the token is unspecified, the whole target field value will be replaced,
+    allowing for replacement of non-string values using the source field type.
+
+#### ObjectReference
+
+A reference to a specific resource object.
+
+-   **apiVersion**: (string)
+
+    The group and version of the object resource.
+
+    One of the following is required: apiVersion or group.
+
+-   **group**: (string)
+
+    The group of the object resource.
+
+    Group is accepted as a version-less alternative to APIVersion.
+
+    Group must be empty for resources in the "core" group.
+
+    One of the following is required: apiVersion or group.
+
+-   **kind**: (string), required
+
+    The kind of the object resource.
+
+-   **name**: (string), required
+
+    The name of the object.
+
+-   **namespace**: (string)
+
+    The namespace of the object.
+
+    Namespace is required for namespaced resources.
+
+<!--mdtogo-->
+
+### Behavior
+
+Like the `depends-on` feature, `apply-time-mutation` waits for dependencies 
+(source resources) to be applied and reconciled before applying the resource
+with the annotation.
+
+Unlike the `depends-on` feature, `apply-time-mutation` modifies the annotated
+resource before applying it.
+
+#### Special cases
+
+If the source resource is not part of the package being applied, the apply of
+the target resource will fail with an error. 
+
+The `apply-time-mutation` annotation is only enforced by `kpt live apply` and
+`kpt live destroy`. Modifying or deleting these resources with other mechanisms
+will not follow the rules specified by these annotations.
+
+### JSONPath syntax
+
+Since there is no formal specification for JSONPath, the supported syntax
+depends on the chosen implimentation. In this case, kpt uses the
+[ajson](https://github.com/spyzhov/ajson) library. For details about what
+language features are supported, see the
+[json-path-comparison table](https://cburgmer.github.io/json-path-comparison/).
+
+### Example
+
+<!--mdtogo:Examples-->
+
+In this example, pod-b depends on pod-a with two substitutions that replace
```

Comment:
The examples here are somewhat more elaborate than the examples for the cli commands and feels almost more like the content of the kpt book. Should we consider making this more like the other examples and reference the book for a more complete description?

Reply:
I wasn't sure how to add apply-time-mutation to the book. And it definitely needs a reference page to specify the full schema.

It sounds like you'd rather have the examples in the book. If so, what should the apply-time-mutation page be called?

FWIW, many of the CLI command examples are actually executable using mdrip for testing. I wasn't sure if we needed that for the annotations, but if we do, it needs a kind cluster to run against.

Reply:
Agree we need a reference page for this and other features/concepts that are too detailed for the book and doesn't fit easily into the cli reference structure. I think this approach looks pretty good for that. I think there are other concepts that we should describe in some more detail outside of the book. Update strategies and function selectors comes to mind.

I think the question is just whether we should leave these things out of the book and then just keep the examples in the reference pages. Since both the depends-on and the apply-time mutations are somewhat advanced features, maybe we don't need to cover it in the book.

For now, I'm ok with keeping this as is and also leave the depends-on chapter in the book. We should probably also reference the apply-time mutations reference page from the book, so anyone following the book can find it.

What do you think @droot ?

I don't think we need to look at running mdrip for these examples in this PR. Automatic verification of the docs is useful, but we should separate look into how to best do it.

Reply:
How about this: 
1. we merge this PR as-is
2. if a customer asks for the apply-time-mutation in the book, or expresses confusion not finding it there, we can add that later.

Reply:
Sounds good